py/runtime.c: Fix op bool for subclasses of native types.

jimmo opened the pull request:
Fixes #5677

This is +8 bytes on PYBV11.

jimmo committed:
py/runtime.c: Fix op bool for subclasses of native types.

Fixes #5677

Previously a subclass of a type that didn't implement unary_op would
raise TypeError on bool conversion.

stinos commented:
I also think this is the correct way to fix it, not 100% sure though. Can you add tests?

jimmo commented:
> I also think this is the correct way to fix it, not 100% sure though. Can you add tests?

Done.

You could also fix this by mandating that all native types have a unary_op with bool implemented, but just doing that for uctypes.struct alone would be about the same code size change. (Interestingly I couldn't find any types in `py/` that would have this issue though).

dpgeorge commented:
This is a neat fix, but I'm not sure it is strictly correct.

The `MP_UNARY_OP_BOOL` operator is supposed to correspond to `__bool__()`.  What's done in this PR is make `mp_unary_op(MP_UNARY_OP_BOOL, obj)` have the semantics of "truth", ie try `__bool__()` then try `__len__()`.

It's a tricky one.  The proper fix is probably to have a function `mp_unary_op_maybe(op, obj)` which is the same as `mp_unary_op(op, obj)` except that it won't raise if the op is unsupported, and instead just return `MP_OBJ_NULL`.  And then use `mp_unary_op` in `instance_unary_op`.  That might add quite a bit of code size (a new function) and maybe impact performance.

Alternative workaround (as pointed out) is to make `uctypes.struct` type implement `MP_UNARY_OP_BOOL`.

jimmo commented:
I've pushed an alternative version (only +4 bytes now).

Relative to the previous version, it no longer falls back to `OP_LEN`, but also no longer explicitly checks for `mp_obj_false`, `mp_obj_true`, `mp_obj_none` (i.e. the first few lines of `mp_obj_is_true`).

False and True are fine, as they implement `.unary_op` (which converts to `OP_BOOL`), which will do the right thing.
So the question is -- can `mp_unary_op(OP_BOOL, x)` ever be called on `x=mp_obj_none`? I can't see any occurences -- the only possible cases are instance_unary_op (can't happen on None), and in the vm directly (OP_BOOL isn't in the bytecode).

stinos commented:
> Alternative workaround (as pointed out) is to make uctypes.struct type implement MP_UNARY_OP_BOOL

This is what I'm doing now for custom types. But if this were chosen as solution maybe `mp_generic_unary_op` should also implement that by default?

dpgeorge commented:
The current fix (testing for `MP_UNARY_OP_BOOL` and returning `mp_const_true` if so) looks OK to me.  At least it fixes the issue and doesn't impact any other cases (?).

For a core type to test this with (rather than uctypes which doesn't exist on all ports), one can use `type`:
```python
class A(type): pass
a = A("foo", (), {})
print(not a)
print(True if a else False)
```
With this PR that test runs the same under CPython as uPy.

jimmo commented:
Rebased, updated the test to use `type` instead of `struct`, and added an explainer comment to the change in mp_unary_op.

stinos commented:
This should fix https://github.com/micropython/micropython/issues/5677 right?

dpgeorge commented on tests/basics/unary_op_builtin.py:
```diff
@@ -0,0 +1,7 @@
+# check user instances derived from builtin without unary op handling for bool.
+class A(type):
```
I think this test should just go in the existing `unary_op.py` file.  There's no longer a reason it needs to be separated (it should be able to run on all uPy ports).

jimmo replied:
Yep, good point. Updated.

jimmo committed:
tests/basics/unary_op.py: unary op bool for native base without unary_op.

This previously (incorrectly) failed with `TypeError: unsupported
type for __bool__: 'type'`.

dpgeorge commented:
Thanks for updating.  Squashed and merged in 4e39ff221abff9d1d7d7fc654da67a89a7543112
